compose/project.py

```diff
@@ -619,46 +621,68 @@ def _get_convergence_plans(self, services, strategy, always_recreate_deps=False)
     def pull(self, service_names=None, ignore_pull_failures=False, parallel_pull=False, silent=False,
              include_deps=False):
         services = self.get_services(service_names, include_deps)
-        msg = not silent and 'Pulling' or None
 
         if parallel_pull:
-            def pull_service(service):
-                strm = service.pull(ignore_pull_failures, True, stream=True)
-                if strm is None:  # Attempting to pull service with no `image` key is a no-op
-                    return
+            self.parallel_pull(services, silent=silent)
 
-                writer = parallel.get_stream_writer()
+        else:
+            for service in services:
+                must_build = []
+                try:
+                    service.pull(ignore_pull_failures, silent=silent)
+                except NotFound:
+                    if service.can_be_built():
+                        must_build.append(service.name)
+                    else:
+                        raise
+
+            if len(must_build):
+                log.warning('Some service image(s) must be built from source by running:\n'
+                            '    docker compose build {}'
+                            .format(' '.join(must_build)))
+
+    def parallel_pull(self, services, ignore_pull_failures=False, silent=False):
+        msg = not silent and 'Pulling' or None
```

Comment:
Not sure if I understood this line, but the following equivalent would be more readable.

```suggestion
        msg = 'Pulling' if not silent else None
```

Reply:
took me time indeed to understand this syntax used in original code.